### ISSUE TITLE
Detections form grid pattern

### ISSUE THREAD

**Issue by graulef** [rpautrat/SuperPoint#79]
Hi Rémi, first of all thanks for this great repo! It really makes it simple to adapt SuperPoint to other tasks and data. 

I am currently trying to apply the model to aerial images taken from a drone. I trained on the standard synthetic shapes for 200'000 iterations and got decent metrics and expected behaviour for the Base Detector (keypoints are located in salient regions of the image, see image below). 

![image](https://user-images.githubusercontent.com/25170019/58968493-6e8c5900-87b6-11e9-9277-3cbc5b737127.png)

I then used this to export pseudo ground-truth on top of my own aerial data and then trained MagicPoint using this for 10'000 iterations with parameters: _detection_threshold_: 0.001, _nms_: 4, _top_k_: 300. 

However, I get the following weird behaviour of the resulting interest point detector: most keypoints seem to be aligned in a grid-like pattern (see image below).

![image](https://user-images.githubusercontent.com/25170019/58967801-23be1180-87b5-11e9-96c8-4ce2e95f082b.png)

If I set _nms_: 8, the "grid distance" increases which seems reasonable.

The loss decreases very quickly while the metrics remain bad: 

```
[05/31/2019 17:35:52 INFO] Start training
[05/31/2019 17:37:49 INFO] Iter    0: loss 4.6257, precision 0.0042, recall 0.0042
[05/31/2019 17:47:20 INFO] Iter 1000: loss 1.7581, precision 0.0077, recall 0.0077
[05/31/2019 17:55:43 INFO] Iter 2000: loss 0.8928, precision 0.0094, recall 0.0094
[05/31/2019 18:04:07 INFO] Iter 3000: loss 0.7964, precision 0.0096, recall 0.0096
[05/31/2019 18:12:25 INFO] Iter 4000: loss 0.7362, precision 0.0123, recall 0.0123
[05/31/2019 18:20:33 INFO] Iter 5000: loss 0.7170, precision 0.0134, recall 0.0134
[05/31/2019 18:28:48 INFO] Iter 6000: loss 0.7516, precision 0.0157, recall 0.0157
[05/31/2019 18:36:58 INFO] Iter 7000: loss 0.7015, precision 0.0152, recall 0.0152
[05/31/2019 18:45:03 INFO] Iter 8000: loss 0.6436, precision 0.0163, recall 0.0163
[05/31/2019 18:53:04 INFO] Iter 9000: loss 0.6845, precision 0.0180, recall 0.0180
[05/31/2019 19:00:41 INFO] Training finished
[05/31/2019 19:00:42 INFO] Saving checkpoint for iteration #10000
```

Have you ever seen similar training behaviour yourself? Do you have an educated guess about what the problem could be here?

My guess is that my aerial dataset is too small, so you might have seen something like this when only training on a subset of COCO or for very few iterations?

**Comment by graulef** [comment]
Also, I have a few more general questions:

1. If I am correctly understand the overall procedure, correctly tuning the detection threshold parameter is crucial, especially when exporting pseudo ground-truth since we need to balance getting too few keypoints vs. reenforcing "random" keypoints. How did you go about when tuning this parameter?

2. How far down does the loss go during training in your experience. And how quickly does it decrease. Do you happen to have a typical loss curve somewhere in your files?

3. Was there some sort of normalization (brightness, histogram or similar) performed on the input images before feeding them to the network?

4. In the original paper, a performance of 70 FPS is reported. Did you ever manage to get similar speeds at inference time?

**Comment by rpautrat** [comment]
Hi!

Regarding your first questions, what is the size of your aerial dataset? It might indeed be a problem if the dataset is too small. Are you training the full model (detector + descriptor) or just MagicPoint (detector only)? The NMS parameter should be adapted to the size of your images as well (I was using a NMS of 4 for 240*320 images). You could also try to train a little bit longer, the metrics were still increasing when you stopped it. But don't pay too much attention to these metrics, they are only a proxy to check the improvement, but the real metrics like repeatability are currently not included in the validation.

For your second set of questions:

1. Indeed, the detection threshold is crucial. I was using 0.001 during the training of SuperPoint to make sure that enough points were detected to match the descriptors. But you should probably use the default value of 0.015 (1/64) during inference to have roughly one detection per patch.

2.This is for example the output of my training of MagicPoint:
```
[10/05/2018 13:26:48 INFO] Start training
[10/05/2018 13:28:05 INFO] Iter    0: loss 4.5987, precision 0.0030, recall 0.0693
[10/05/2018 13:45:56 INFO] Iter 1000: loss 1.4893, precision 0.0037, recall 0.0766
[10/05/2018 14:03:31 INFO] Iter 2000: loss 0.7217, precision 0.0043, recall 0.0920
[10/05/2018 14:15:44 INFO] Iter 3000: loss 0.5048, precision 0.0046, recall 0.0986
[10/05/2018 14:23:54 INFO] Iter 4000: loss 0.4139, precision 0.0061, recall 0.1117
[10/05/2018 14:31:03 INFO] Iter 5000: loss 0.4702, precision 0.0170, recall 0.1145
[10/05/2018 14:38:08 INFO] Iter 6000: loss 0.3807, precision 0.0219, recall 0.1239
[10/05/2018 14:44:53 INFO] Iter 7000: loss 0.3890, precision 0.0251, recall 0.1262
[10/05/2018 14:51:38 INFO] Iter 8000: loss 0.4116, precision 0.0258, recall 0.1242
[10/05/2018 14:58:22 INFO] Iter 9000: loss 0.3463, precision 0.0310, recall 0.1301
[10/05/2018 15:05:30 INFO] Iter 10000: loss 0.3407, precision 0.0306, recall 0.1338
[10/05/2018 15:12:38 INFO] Iter 11000: loss 0.3442, precision 0.0334, recall 0.1378
[10/05/2018 15:19:24 INFO] Iter 12000: loss 0.3531, precision 0.0343, recall 0.1367
[10/05/2018 15:26:09 INFO] Iter 13000: loss 0.3976, precision 0.0300, recall 0.1352
[10/05/2018 15:32:55 INFO] Iter 14000: loss 0.3376, precision 0.0331, recall 0.1339
[10/05/2018 15:39:43 INFO] Iter 15000: loss 0.3185, precision 0.0315, recall 0.1350
[10/05/2018 15:46:29 INFO] Iter 16000: loss 0.3793, precision 0.0328, recall 0.1434
[10/05/2018 15:53:11 INFO] Iter 17000: loss 0.3996, precision 0.0297, recall 0.1367
[10/05/2018 15:59:46 INFO] Training finished
[10/05/2018 15:59:51 INFO] Saving checkpoint for iteration #18000
```
So you could probably get a better loss and metrics on your own dataset than what you currently have.

3. There is no prior normalization of the images (except converting them to grayscale).

4. No, I never got such a speed at inference time. For the full network, I think that I got more something like 3 FPS... But the the network could be easily compressed (for example replacing the VGG by something lighter, like a MobileNet) without a significant loss in performance.

**Comment by graulef** [comment]
Thanks a lot for your answers!

The dataset is very small (~200 MB) so I will probably add some images from COCO as well and train longer.

In regards to your answer to 1: so if we set the detection threshold to 1/65 when generating pseudo ground-truth, I understand that we only detect one point per warped image. So if we set the number of homographic adaptions to 100, we would get 100 detections. So would the parameter `top_k` be more or less irrelevant during this step? Is that also why its commented out in your config file?
And then what about the actual training of MagicPoint, would you still leave it at 1/65 or suggest a lower value then?

Thanks as well for the other answers, they are very helpful. Highly appreciate how active this repo is!

**Comment by rpautrat** [comment]
Hi,

I used the threshold of 0.015 only to create the pseudo ground truth and 0,001 to actually train MagicPoint.

0,015 doesn't mean that there will be one point detected per warped image, but rather one point per 8x8 patch (the softmax is performed along the channel axis, i.e. within each small patch). So if your image is WxH, you will have on average W/8xH/8 points detected in an image.

In the end I didn't use the parameter top_k during training, specifying the right threshold was enough to control the amount of points detected.
